Add full image support for UniElec U7623

dwmw2 opened the pull request:
This adds a full eMMC image for the UniElec U7623, including U-Boot, and renames the existing image to "legacy".

Sysupgrade is possible from 19.07 using the "legacy" image, and from there to the full image.

It also adds 'scatter files' for the mt7623 targets, for unbricking devices with the MediaTek SP Flash Tool.

dwmw2 committed:
mediatek: fix Ethernet support on UniElec U7623

The definition of the switch in the device-tree was not correct. Make it
look more like the Banana Pi R2, which works.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: add U-Boot build for UniElec U7623

Patches submitted upstream at
https://patchwork.ozlabs.org/project/uboot/list/?series=189178

Tested on Banana Pi R2 and U7623-06.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: increase flexibility of finding recovery partition

I'm about to change the layout of the images for UniElec U7623 so make it
find the recovery partition based on which the root partition is too.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: add explicit console= to U7623 kernel

The bootloader for legacy builds can't set it, so we end up unable to
log in on the serial port.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: prepare for full sysupgrade support on UniElec U7623

An upcoming commit will add a full system image for U7623 which will
contain the MBR partition table and U-Boot too.

That contrasts with the current image which only owns the eMMC from
sector 0xa00 onwards, and must start with a legacy uImage.

Prepare for sysupgrade to the new images, and cope with the fact that
the recovery partition will be /dev/mmcblk0p2 instead of /dev/mmcblk0p1
after the upgrade.

This commit could potentially be backported to 19.07 to allow for direct
sysupgrade to the new image layout.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
@kristrev FYI

adrianschmutzler commented on package/boot/uboot-mediatek/Makefile:
```diff
@@ -34,7 +34,14 @@ define U-Boot/mt7623n_bpir2
 endef
 
-UBOOT_TARGETS := mt7629 mt7622 mt7623n_bpir2
+define U-Boot/mt7623a_unielec_u7623
```
This file could need some sorting. I'd be happy if you just moved the "devices" according to sorting in this patch as well.

Same for the targets line below.

dwmw2 replied:
Done.

adrianschmutzler commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -42,16 +42,21 @@ preinit_set_mac_address() {
 
 	case $(board_name) in
-	bananapi,bpi-r2)
+	bananapi,bpi-r2|unielec,u7623-02-emmc-512m)
```
I'd prefer wrapped version:

```
	bananapi,bpi-r2|\
	unielec,u7623-02-emmc-512m)
```

dwmw2 replied:
Done.

adrianschmutzler commented on target/linux/mediatek/image/gen_scatterfile.sh:
```diff
@@ -0,0 +1,77 @@
+  reserve: 0x00
+EOF
+
+
```
are the empty lines at EOF needed (I have no idea what this file actually does)?

dwmw2 replied:
It's just a shell script; there's no need for the empty lines. Removed.

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -9,22 +9,34 @@ define Build/banana-pi-sdcard
+
+define Build/scatterfile
+	./gen_scatterfile.sh $(subst mt,MT,$(SUBTARGET)) "$1" \
+		$(subst -scatter.txt,,$(notdir $@)) "$(DEVICE_VENDOR) $(DEVICE_MODEL)" > $@
```
Is there a specific reason for using only DEVICE_VENDOR and DEVICE_MODEL here manually, and not just using DEVICE_TITLE that is assembled from that + DEVICE_VARIANT?

dwmw2 replied:
No reason; it's purely cosmetic anyway, to give the user a clue when they read the scatter text file (which is a list of regions given to the USB debricking tool). I can change it.

adrianschmutzler replied:
If it's cosmetic, I'd go for the standard/appropriate variable which is DEVICE_TITLE:
https://github.com/openwrt/openwrt/blob/263f7e5bbd119ebed1f514c16f659a2e2a2b132c/include/image.mk#L360

I'd only use something different for a reason.

dwmw2 replied:
Fixed; thanks.

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -33,21 +45,53 @@ define Device/bpi_bananapi-r2
 	mkf2fs e2fsprogs kmod-usb-ohci kmod-usb2 kmod-usb3 kmod-ata-ahci-mtk
   UBOOT_TARGET := mt7623n_bpir2
+  UBOOT_OFFSET := 320k
+  UBOOT_ENVSIZE := 0x2000
```
Are these added to DEVICE_VARS somewhere? (Haven't checked myself)

dwmw2 replied:
They are now.

dwmw2 committed:
mediatek: mt7623: fix sysupgrade from vendor OpenWrt on UniElec U7623

This board ships with an ancient 14.07-based OpenWrt using block2mtd, and
the MBR partition table contains nonsense.

It is possible to sysupgrade to an upstream OpenWrt image, but the
legacy layout of the OpenWrt images start at 0xA00 in the eMMC, with
a raw uImage. The legacy OpenWrt image doesn't "own" the beginning
of the device, including the MBR and U-Boot.

This means that when a user upgrades to upstream OpenWrt, it doesn't
boot because it can't find the right partitions. So hard-code them on
the kernel's command line using CONFIG_CMDLINE_PARTITION (for block).

Additionally, the vendor firmware doesn't cope with images larger than
about 36MiB, because it only overwrites the contents of its "firmware"
MTD partition. The current layout of the legacy image wastes a lot of
space, allowing over 32MiB for the kernel and another 10MiB for the FAT
recovery file system which is only created as 3MiB. So pull those in
to allow 4¾ MiB for the kernel, 3MiB for recovery, and then we have over
20MiB for the root file system.

This doesn't affect the new images which ship with a full eMMC image
including a different MBR layout and a partition for U-Boot, because
our modern U-Boot can actually pass the command line to the kernel, and
the built-in one doesn't get used anyway.

Tested by upgrading from vendor OpenWrt to the current legacy image,
from legacy to itself, to the previous legacy layout, and then to
finally the full-system image.

This commit probably wants backporting to 19.07, which also doesn't
install over the vendor OpenWrt and doesn't even have a full-system
installation option.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

mans0n commented on target/linux/mediatek/image/Makefile:
```diff
@@ -29,8 +29,8 @@ define Device/Default
-  KERNEL := kernel-bin | append-dtb | uImage none
-  KERNEL_INITRAMFS := kernel-bin | append-dtb | uImage none
+  KERNEL = kernel-bin | fit none $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  KERNEL_INITRAMFS :=
```
If you are going to touch this, then why do you add the same lines in device recipes?
BTW why do you disable initramfs images? It can still be used for debugging purpose.

dwmw2 replied:
It's only the legacy U7623 recipe which gets this. The others could even use `fit lzma` like everything else but there's no point in compression a since it's a zImage so I left it with "none".

dwmw2 replied:
I put the initramfs image back.

mans0n commented:
IMO the last commit is quite big (you change a lot of things), and it's hard to track what is changed and what is not... Could you split it into smaller logical blocks? (I guess scatterfile or metadata can be in a separate patch?)

dwmw2 committed:
mediatek: mt7623: change default kernel build to FIT image

Currently there are only two targets, one of which (BPi R2) uses FIT and
the other (UniElec U7623) a legacy uImage. But I'm about to add a new
image for U7623 with a modern U-Boot which uses FIT images too.

So switch the default for MT7623 to be FIT, and let the legacy U7623
target override it to uImage, instead of the converse.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: add scatter file for unbricking with SP Flash Tool

Many MediaTek SoCs can be unbricked by using the SP Flash Tool from
http://spflashtool.com/ along with a "scatter list" file, which is
just a text file listing which image gets loaded where.

We use a trivial partition layout for the tool, with the whole eMMC
image as a single "partition", which means users just need to unzip
the sysupgrade image. Doing the real partition layout would be overly
complex and would require the individual partitions to be shipped
as artifacts — or users to extract them out of the sysupgrade image
just for the tool to put them adjacent to each other on the eMMC
anyway.

The tool does require a copy of the preloader in order to operate,
even when it isn't flashing the preloader to the eMMC boot region.
So drop that into the bin directory as an artifact too.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: make gen_banana_pi_img.sh more generic

This actually covers fairly much all the MediaTek platforms; they
only have different images because they don't include the preloader
and U-Boot, and rely on preinstalled stuff from the vendor.

So this script can slowly take over the world as we complete the
support for various other platforms, starting with UniElec U7623…

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: rename gen_banana_pi_img.sh → gen_mtk_mmc_img.sh

As I buy more hardware and continue to work on consolidation, This will
apply to a lot of MediaTek platforms; rename it accordingly.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 committed:
mediatek: mt7623: add full system image for UniElec U7623

This adds a full eMMC image including U-Boot, which means that the
kernel can inherit the true RAM size detected by the preloader.

As implemented in previous commits, sysupgrade to this image from
the legacy layout (and via that, from the vendor-installed image)
is supported.

Rename the legacy image for the 512MiB board, for clarity.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

dwmw2 commented:
> IMO the last commit is quite big (you change a lot of things), and it's hard to track what is changed and what is not... Could you split it into smaller logical blocks? (Perhaps scatterfile or metadata can be in a separate patch?)

Now split.

adrianschmutzler commented on target/linux/mediatek/image/Makefile:
```diff
@@ -27,15 +27,9 @@ define Device/Default
   IMAGE/sysupgrade.bin := append-kernel | pad-to 128k | append-rootfs | pad-rootfs | append-metadata
   SUPPORTED_DEVICES := $(subst _,$(comma),$(1))
-ifeq ($(SUBTARGET),mt7623)
-  KERNEL_NAME := zImage
```
Please also remove the redundant unused KERNEL_NAME six lines above.

dwmw2 replied:
done.

adrianschmutzler commented on target/linux/mediatek/image/mt7623.mk:
```diff
@@ -1,53 +1,95 @@
+DEVICE_VARS += UBOOT_TARGET UBOOT_OFFSET UBOOT_ENVSIZE
 
-ifneq ($(CONFIG_BANANA_PI_BOOT_PARTSIZE),)
+ifneq ($(CONFIG_MTK_BOOT_PARTSIZE),)
```
Most other target add "SD" here:
CONFIG_AXS10X_SD_BOOT_PARTSIZE
CONFIG_OMAP_SD_BOOT_PARTSIZE
AT91_SD_BOOT_PARTSIZE
CONFIG_SUNXI_SD_BOOT_PARTSIZE

In contrast, I wonder whether the target name needs to be included here, as it's defined per target anyway. Adding the device name would make sense if there are different values for different targets, but if it's the same across the target the target name won't provide any information.

dwmw2 replied:
I think it had SD in it once, but it covers eMMC too so I figured that was misleading.

adrianschmutzler replied:
err, you are right, I'm too much consumer there, for me it's all "SD" already ...

dwmw2 commented:
#3237 is the minimal fixes for 19.07 to at least allow 19.07 to be installed from the vendor firmware.

I'll probably also backport the sysupgrade bits so that users can upgrade directly to the full (OpenWrt 20.xx) image from 19.07 without having to go via the legacy image.

dwmw2 committed:
mediatek: mt7623: fix Ethernet setup for TRGMII mode

This fixes the TX performance issues seen on MT7623 boards.

Signed-off-by: David Woodhouse <dwmw2@infradead.org>

981213 reviewed (commented):
Hi!

The following build error occurs due to missing mt7623a_unielec_u7623-uEnv.txt:
```
mkenvimage -s 0x1000 -o /mnt/satadrive/owrt/master/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/image/mt7623a_unielec_u7623-uboot.env mt7623a_unielec_u7623-uEnv.txt
Can't open "mt7623a_unielec_u7623-uEnv.txt": No such file or directory
make[4]: *** [Makefile:42: /mnt/satadrive/owrt/master/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-mediatek_mt7623/tmp/openwrt-mediatek-mt7623-unielec_u7623-emmc-squashfs-generic.img.gz] Error 1
```

and one typo below:

dwmw2 commented on target/linux/mediatek/mt7623/base-files/lib/preinit/07_set_iface_mac:
```diff
@@ -42,16 +42,22 @@ preinit_set_mac_address() {
-		set_recovery_mac_address /dev/mmcblk0p1
+		rootpart="${rootpart%% *}"
+		case $rootpart in
+		/dev/mmcblk0p2) # Legacy U7632 image
```
Fixed; thanks.

981213 replied:
while at it, this typo is still here :)

dwmw2 commented:
> The following build error occurs due to missing mt7623a_unielec_u7623-uEnv.txt:

Fixed; thanks.

dwmw2 commented:
There is pointless churn in the tree just making this more complex. Closing for now. Let me know if you're interested in contributions.

981213 commented:
Whoops.
Would you mind to rebase it? I'll merge this after a rebase.

dwmw2 commented:
>  Would you mind to rebase it? I'll merge this after a rebase.

OK, one more update for this and the corresponding #3237 for 19.07.

981213 commented:
I squashed
Revert "mediatek: remove condition in Device/Default"
mediatek: mt7623: change default kernel build to FIT image
together and reworked the commit message, resulting in the following commit:
https://git.openwrt.org/65fc47cac562f0bba18aafd40e2d51884075e5ab
after resolving two conflicts in later commits, the final result is merged into my staging tree:
https://git.openwrt.org/?p=openwrt/staging/981213.git
I'll push it to the main tree after a build test. Thanks!

dwmw2 commented:
I was happier with explicitly reverting the gratuitously conflicting commit that I had asked not to happen. But OK. Thanks.

dwmw2 commented:
If you do push your variant please also remember to change commit IDs in the 'cherry picked from..." in #3237. I had fixed those up already this morning.

adrianschmutzler commented:
> There is pointless churn in the tree just making this more complex. Closing for now. Let me know if you're interested in contributions.

From my experience, the typical waiting time for a commit of this size is between 2 and 6 months (some even don't get merged at all).

I have merged a change that essentially did part of what you tried to achieve here anyway, and you just would have had to do a relatively simple rebase.

So, as I could not expect you to be treated special, anticipating the above stated waiting time, it made sense to merge my commit right away.

This whole PR has been open for only 6 days; there are a lot of PRs that don't even receive initial feedback within that timespan. So, you can consider yourself really lucky that you receive this very special treatment of getting merged almost instantly, while most of the other people having invested work (and money) into their contributions as well still wait for somebody to look at their stuff.

dwmw2 commented:
Sure, lots of projects have issues with maintainer bandwidth and in the worst cases potential contributors do indeed have to wait for months even for a PR which is already ready to go. I was under no illusions about that. I'd actually taken over some of the Banana Pi R2 work from a potential contributor who'd just given up and didn't even want to bother. I can cope with that.

What offended me was the purely gratuitous cosmetic commit, making no functional improvement, which was completely redundant with the cleanup I had made myself. And which I had asked to be dropped.

To me, that seemed just plain rude.

adrianschmutzler commented:
> And which I had asked to be dropped.

Your sole comment was "#3230 now unifies this anyway.". I don't think that's enough for justifying your level of offense.

dwmw2 commented:
Oh, come on. Take a look at the actual difference now the churn has been resolved. It's a few whitespace changes and a few lines moved up/down (but overall not a lot more sorted than before I believe). It's an utterly gratuitous merge conflict.

What did you *think* my point was, when I said those words?

dwmw2 commented:
Anyway, let's not lose sleep over it.

It was a suboptimal way to handle a contribution, and my grumpy reaction was also suboptimal. Apologies for my part in that.

The important point is that the boards in question are better supported now, and there are more similar boards on my pile waiting for the same treatment. If I am a bit slow in continuing with the mt7622 boards it's because real work and real life have intervened; not because I've thrown my toys out of the pram :)
